File: backend/src/main/java/com/biznest/backend/service/ReviewService.java
package com.biznest.backend.service;

import com.biznest.backend.model.Review;
import com.biznest.backend.model.BusinessListing;
import com.biznest.backend.model.UserEntity;
import com.biznest.backend.repository.ReviewRepository;
import com.biznest.backend.repository.BusinessListingRepository;
import com.biznest.backend.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class ReviewService {
    
    @Autowired
    private ReviewRepository reviewRepository;
    
    @Autowired
    private BusinessListingRepository businessListingRepository;
    
    @Autowired
    private UserRepository userRepository;

    public Review createReview(Review review, Long userId, Long businessListingId) {
        UserEntity user = userRepository.findById(userId)
            .orElseThrow(() -> new RuntimeException("User not found"));
        BusinessListing businessListing = businessListingRepository.findById(businessListingId)
            .orElseThrow(() -> new RuntimeException("Business listing not found"));
        
        // Check if user has already reviewed this business
        if (reviewRepository.existsByUserAndBusinessListing(user, businessListing)) {
            throw new RuntimeException("User has already reviewed this business");
        }
        
        review.setUser(user);
        review.setBusinessListing(businessListing);
        
        Review savedReview = reviewRepository.save(review);
        
        // Update business listing average rating
        updateBusinessListingRating(businessListing);
        
        return savedReview;
    }

    public Optional<Review> getReview(Long id) {
        return reviewRepository.findById(id);
    }

    public List<Review> getReviewsByBusinessListing(Long businessListingId) {
        BusinessListing businessListing = businessListingRepository.findById(businessListingId)
            .orElseThrow(() -> new RuntimeException("Business listing not found"));
        return reviewRepository.findByBusinessListingOrderByCreatedAtDesc(businessListing);
    }

    public List<Review> getReviewsByUser(Long userId) {
        UserEntity user = userRepository.findById(userId)
            .orElseThrow(() -> new RuntimeException("User not found"));
        return reviewRepository.findByUserOrderByCreatedAtDesc(user);
    }

    public Page<Review> getReviewsByBusinessListing(Long businessListingId, Pageable pageable) {
        BusinessListing businessListing = businessListingRepository.findById(businessListingId)
            .orElseThrow(() -> new RuntimeException("Business listing not found"));
        return reviewRepository.findByBusinessListing(businessListing, pageable);
    }

    public Page<Review> getReviewsByUser(Long userId, Pageable pageable) {
        UserEntity user = userRepository.findById(userId)
            .orElseThrow(() -> new RuntimeException("User not found"));
        return reviewRepository.findByUser(user, pageable);
    }

    public Review updateReview(Long id, Review updatedReview) {
        Review existing = reviewRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Review not found"));
        
        existing.setRating(updatedReview.getRating());
        existing.setComment(updatedReview.getComment());
        Review savedReview = reviewRepository.save(existing);
        
        // Update business listing average rating after review update
        updateBusinessListingRating(existing.getBusinessListing());
        
        return savedReview;
    }

    public Review addBusinessResponse(Long reviewId, String response) {
        return reviewRepository.findById(reviewId)
            .map(existing -> {
                existing.setBusinessResponse(response);
                return reviewRepository.save(existing);
            })
            .orElseThrow(() -> new RuntimeException("Review not found"));
    }

    public void deleteReview(Long id) {
        Review review = reviewRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Review not found"));
        
        BusinessListing businessListing = review.getBusinessListing();
        reviewRepository.deleteById(id);
        
        // Update business listing average rating after deletion
        updateBusinessListingRating(businessListing);
    }

    public Double getAverageRatingByBusinessListing(Long businessListingId) {
        BusinessListing businessListing = businessListingRepository.findById(businessListingId)
            .orElseThrow(() -> new RuntimeException("Business listing not found"));
        return reviewRepository.getAverageRatingByBusinessListing(businessListing);
    }

    public Long getReviewCountByBusinessListing(Long businessListingId) {
        BusinessListing businessListing = businessListingRepository.findById(businessListingId)
            .orElseThrow(() -> new RuntimeException("Business listing not found"));
        return reviewRepository.getReviewCountByBusinessListing(businessListing);
    }

    public List<Review> getReviewsWithBusinessResponse() {
        return reviewRepository.findByBusinessResponseIsNotNull();
    }

    public List<Review> getReviewsWithoutBusinessResponse() {
        return reviewRepository.findByBusinessResponseIsNull();
    }

    public boolean hasUserReviewedBusiness(Long userId, Long businessListingId) {
        UserEntity user = userRepository.findById(userId)
            .orElseThrow(() -> new RuntimeException("User not found"));
        BusinessListing businessListing = businessListingRepository.findById(businessListingId)
            .orElseThrow(() -> new RuntimeException("Business listing not found"));
        return reviewRepository.existsByUserAndBusinessListing(user, businessListing);
    }

    private void updateBusinessListingRating(BusinessListing businessListing) {
        Double averageRating = reviewRepository.getAverageRatingByBusinessListing(businessListing);
        if (averageRating != null) {
            // Round to 1 decimal place
            double roundedRating = Math.round(averageRating * 10.0) / 10.0;
            businessListing.setRating(roundedRating);
            businessListingRepository.save(businessListing);
        }
    }
} 

File: frontend/src/services/reviewService.js
import api from "./api";

const reviewService = {
  // Create a new review
  createReview: async (review, userId, businessListingId) => {
    const params = new URLSearchParams();
    params.append('userId', userId);
    params.append('businessListingId', businessListingId);
    const response = await api.post(`/api/reviews?${params.toString()}`, review);
    return response.data;
  },

  // Get a single review by ID
  getReview: async (id) => {
    const response = await api.get(`/api/reviews/${id}`);
    return response.data;
  },

  // Get all reviews for a business listing
  getReviewsByBusinessListing: async (businessListingId) => {
    const response = await api.get(`/api/reviews/business/${businessListingId}`);
    return response.data;
  },

  // Get all reviews for a business listing (paginated)
  getReviewsByBusinessListingPaginated: async (businessListingId, page = 0, size = 10) => {
    const response = await api.get(`/api/reviews/business/${businessListingId}/page?page=${page}&size=${size}`);
    return response.data;
  },

  // Get all reviews by a user
  getReviewsByUser: async (userId) => {
    const response = await api.get(`/api/reviews/user/${userId}`);
    return response.data;
  },

  // Get all reviews by a user (paginated)
  getReviewsByUserPaginated: async (userId, page = 0, size = 10) => {
    const response = await api.get(`/api/reviews/user/${userId}/page?page=${page}&size=${size}`);
    return response.data;
  },

  // Update a review
  updateReview: async (id, updated) => {
    const response = await api.put(`/api/reviews/${id}`, updated);
    return response.data;
  },

  // Add business response to a review
  addBusinessResponse: async (id, response) => {
    const responseData = await api.post(`/api/reviews/${id}/response`, response);
    return responseData.data;
  },

  // Delete a review
  deleteReview: async (id) => {
    const response = await api.delete(`/api/reviews/${id}`);
    return response.data;
  },

  // Get average rating for a business listing
  getAverageRating: async (businessListingId) => {
    const response = await api.get(`/api/reviews/business/${businessListingId}/average-rating`);
    return response.data;
  },

  // Get review count for a business listing
  getReviewCount: async (businessListingId) => {
    const response = await api.get(`/api/reviews/business/${businessListingId}/count`);
    return response.data;
  },

  // Get reviews with business response
  getReviewsWithBusinessResponse: async () => {
    const response = await api.get('/api/reviews/with-response');
    return response.data;
  },

  // Get reviews without business response
  getReviewsWithoutBusinessResponse: async () => {
    const response = await api.get('/api/reviews/without-response');
    return response.data;
  },

  // Check if user has reviewed a business
  hasUserReviewedBusiness: async (userId, businessListingId) => {
    const params = new URLSearchParams();
    params.append('userId', userId);
    params.append('businessListingId', businessListingId);
    const response = await api.get(`/api/reviews/check-reviewed?${params.toString()}`);
    return response.data;
  }
};

export default reviewService; 